### PR TITLE
Add cluster override and handle case of write protected DBFS

### PR DESCRIPTION
**Pull request by dmoore247**
May fix https://github.com/databrickslabs/ucx/issues/493

### CODE REVIEW

**Comment by dmoore247** [comment]
Acceptance test was good except two timeout errors

**Comment by nfx** [comment]
@dmoore247 can you run those two acceptance tests locally and see if they pass. attach a screenshot to confirm

**Comment by dmoore247** [comment]
@nfx The timeout errors fail on local test, the testing framework breaks.
```
>       ws_group_a, acc_group_a = make_ucx_group()

tests/integration/test_installation.py:78: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/integration/conftest.py:97: in inner
    members = [_.id for _ in random.choices(user_pool, k=random.randint(1, 40))]
../../.pyenv/versions/3.10.10/lib/python3.10/random.py:519: in choices
    return [population[floor(random() * n)] for i in _repeat(None, k)]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = repeat(None, 27)

>   return [population[floor(random() * n)] for i in _repeat(None, k)]
E   IndexError: list index out of range

../../.pyenv/versions/3.10.10/lib/python3.10/random.py:519: IndexError
```
```
@pytest.fixture
def make_ucx_group(make_random, make_group, make_acc_group, user_pool):
    def inner():
        display_name = f"ucx_{make_random(4)}"
        members = [_.id for _ in random.choices(user_pool, k=random.randint(1, 40))]
        ws_group = make_group(display_name=display_name, members=members, entitlements=["allow-cluster-create"])
        acc_group = make_acc_group(display_name=display_name, members=members)
        return ws_group, acc_group

    return inner
    ```

**Comment by dmoore247** [comment]
Local integration test results...
`test_jobs_with_no_inventory_database` failure due to #615 

For test `test_set_owner_permission` noticed that any group operation is extremely slow, even through UI, command line.

image

**Review by renardeinside — CHANGES_REQUESTED** [review]
please update from latest main first

**Comment by dmoore247** [comment]
Integration tests ok except for 3 timeouts...
```
FAILED tests/integration/workspace_access/test_groups.py::test_set_owner_permission - TimeoutError: Timed out after 0:15:00
FAILED tests/integration/workspace_access/test_workspace_access.py::test_workspace_access_e2e - TimeoutError: Timed out after 0:15:00
FAILED tests/integration/test_installation.py::test_jobs_with_no_inventory_database - TimeoutError: Timed out after 0:25:00
```

**Comment by dmoore247** [comment]
@renardeinside I updated branch from main. Please ok this.

**Comment by dmoore247** [comment]
@renardeinside I updated from Main again, 2nd time. Please approve before this ages out again.

**Comment by dmoore247** [comment]
@nfx 3rd time I've had to refresh this branch (CI build still running).
Can you take a look and merge this soonest please?

**Comment by dmoore247** [comment]
@nfx Integration tests perform as expected
image

**Comment by dmoore247** [comment]
```
=========================== short test summary info ============================
FAILED tests/integration/workspace_access/test_groups.py::test_set_owner_permission - TimeoutError: Timed out after 0:15:00
FAILED tests/integration/test_installation.py::test_jobs_with_no_inventory_database - TimeoutError: Timed out after 0:25:00
============= 2 failed, 49 passed, 6 skipped in 1633.84s (0:27:13) =============

File: src/databricks/labs/ucx/install.py [hunk 1]
```diff
@@ -343,6 +345,34 @@ def _configure_inventory_database(self):
...
 
+    def _valid_cluster_id(self, cluster_id: str) -> bool:
+        """Lite validation of user supplied cluster id"""
+        return CLUSTER_ID_LENGTH == len(cluster_id)
+
+    def _configure_override_clusters(self):
+        """User may override standard job clusters with interactive clusters"""
+        default_val = ""
+        preamble = "Pick default and we'll create a compatible job cluster (recommended) or "
+        cluster_id = self._question(
```
Comment (FastLee) [inline src/databricks/labs/ucx/install.py:356]: Use _prompt and present the user with a list of interactive clusters (maybe up to 20)

File: src/databricks/labs/ucx/install.py [hunk 2]
```diff
@@ -343,6 +345,34 @@ def _configure_inventory_database(self):
                     raise SystemExit(msg)
         return inventory_database
 
+    def _valid_cluster_id(self, cluster_id: str) -> bool:
```
Comment (FastLee) [inline src/databricks/labs/ucx/install.py:348]: Validate the cluster name exists.
Use the warehouse selection as a reference.

File: src/databricks/labs/ucx/install.py [hunk 3]
```diff
@@ -622,9 +654,20 @@ def _upload_wheel(self) -> str:
...
             remote_dirname = os.path.dirname(remote_wheel)
             with local_wheel.open("rb") as f:
-                self._ws.dbfs.mkdirs(remote_dirname)
-                logger.info(f"Uploading wheel to dbfs:{remote_wheel}")
-                self._ws.dbfs.upload(remote_wheel, f, overwrite=True)
+                try:
+                    self._ws.dbfs.mkdirs(remote_dirname)
+                    logger.info(f"Uploading wheel to dbfs:{remote_wheel}")
+                    self._ws.dbfs.upload(remote_wheel, f, overwrite=True)
+                except OperationFailed as err:
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:661]: this is an incorrect exception to catch - it's not thrown in the `dbfs.upload()` code path. 

you have to catch `databricks.sdk.errors.mapping.PermissionDenied` (i think) - we don't have an environment to reproduce this.

File: src/databricks/labs/ucx/install.py [hunk 4]
```diff
@@ -622,9 +654,20 @@ def _upload_wheel(self) -> str:
...
+                    logger.warning(f"Uploading wheel file to DBFS failed, DBFS is probably write protected. {err}")
+                    # assume DBFS is write protected
+                    self._write_protected_dbfs = True
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:664]: can we ask for `self._configure_override_clusters()` in this branch? it'll be silly to make installation more complex for 99% of customers, who don't have a write-protected DBFS.

**Review by nfx — CHANGES_REQUESTED** [review]
Please make the cluster override prompt shown to the user only when write protected DBFS is detected. No need to make installation more complex for 99% of the customers, who don't have this problem.

`OperationFailed` is an incorrect exception to catch - it's not thrown in the dbfs.upload() code path. you have to catch databricks.sdk.errors.mapping.PermissionDenied (i think) - please double-check, as we don't have an environment to reproduce this.

It's also not clear for the user what does the `Pick default and we'll create a compatible job cluster (recommended)` mean. 

- what if user selects the same cluster for both? 
- what if user selects a non-tacl cluster for tacl work? 
- can you add more validation here?
- what will happen if user runs `databricks labs upgrade ucx`? where the cluster overrides would be coming from then? i think there's a need to put it into a configuration file as well.
- i think we still upload a wheel even for cluster override tasks configurations - see https://github.com/databrickslabs/ucx/blob/main/src/databricks/labs/ucx/install.py#L675-L685

File: src/databricks/labs/ucx/install.py (hunk 4 above)
Comment (dmoore247) [reply src/databricks/labs/ucx/install.py:664]: Hmmm... cluster overrides are handy for a lot of cases > 1%, but not more than 50%.

Comment (dmoore247) [reply src/databricks/labs/ucx/install.py:664]: The intended select logic will return only clusters with the correct security mode.

File: src/databricks/labs/ucx/install.py [hunk 5]
```diff
@@ -622,9 +654,20 @@ def _upload_wheel(self) -> str:
...
+                    if not self._override_clusters:
+                        logger.error("Recommend adding override clusters with attached wheel file as python library.")
```
Comment (dmoore247) [inline src/databricks/labs/ucx/install.py:666]: This could be a good place to trigger the configure_override_clusters

Comment (dmoore247) [reply src/databricks/labs/ucx/install.py]: Done

File: src/databricks/labs/ucx/install.py (hunk 4 above)
Comment (dmoore247) [reply src/databricks/labs/ucx/install.py:642]: Done

File: src/databricks/labs/ucx/install.py (hunk 3 above)
Comment (dmoore247) [reply src/databricks/labs/ucx/install.py]: Just pushed this.

File: src/databricks/labs/ucx/install.py (hunk 5 above)
Comment (dmoore247) [reply src/databricks/labs/ucx/install.py]: Done, configuring on DBFS PermissionDenied exception

File: src/databricks/labs/ucx/install.py (hunk 4 above)
Comment (dmoore247) [reply src/databricks/labs/ucx/install.py:641]: Done

Comment (dmoore247) [reply src/databricks/labs/ucx/install.py:641]: @nfx The `config.yml` file is written before the upload wheel is attempted and the write protected DBFS is detected. Delaying the cluster config means that the overrides are not available when it comes time to write the config. Do we write twice, do we delay the write of the config, does that mess up other logic?

**Comment by dmoore247** [comment]
Two failing integration tests were run locally.

One has been failing for months and is documented with bug report: https://github.com/databrickslabs/ucx/issues/663 

The other ran locally.
image

File: src/databricks/labs/ucx/configure.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+        def build_and_prompt(prompt, clusters):
+            choices = {"[Default, use a job cluster]": None}
+            for _ in clusters:
+                choices[_.cluster_name] = _.cluster_id
+            return self._choice_from_dict(prompt, choices=choices)
+
+        clusters = self._ws.clusters.list(can_use_client="NOTEBOOK")
+        classic_clusters = [c for c in clusters if is_classic(c)]
+        tacl_clusters = [c for c in clusters if is_tacl(c)]
```
Comment (nfx) [inline src/databricks/labs/ucx/configure.py:46]: please rewrite lines 44-46 as a single for loop. it was causing a bug is GroupManager before

File: src/databricks/labs/ucx/install.py [hunk 7]
```diff
@@ -623,9 +631,18 @@ def _upload_wheel(self) -> str:
...
-                logger.info(f"Uploading wheel to dbfs:{remote_wheel}")
-                self._ws.dbfs.upload(remote_wheel, f, overwrite=True)
+                try:
+                    self._ws.dbfs.mkdirs(remote_dirname)
+                    logger.info(f"Uploading wheel to dbfs:{remote_wheel}")
+                    self._ws.dbfs.upload(remote_wheel, f, overwrite=True)
+                except PermissionDenied as err:
+                    logger.warning(f"Uploading wheel file to DBFS failed, DBFS is probably write protected. {err}")
+                    # assume DBFS is write protected
+                    self._write_protected_dbfs = True
```
Comment (nfx) [inline src/databricks/labs/ucx/install.py:641]: `._write_protected_dbfs` doesn't seem to be used anywhere

File: src/databricks/labs/ucx/configure.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+from databricks.sdk.service import compute
+
+logger = logging.getLogger(__name__)
+CLUSTER_ID_LENGTH = 20  # number of characters in a valid cluster_id
+
+
+class ConfigureClusterOverrides:
+    """Installation configuration operations to suplement install.WorkspaceInstaller"""
+
+    def __init__(self, ws, _choice_from_dict):
```
Comment (nfx) [inline src/databricks/labs/ucx/configure.py:12]:
```suggestion
    def __init__(self, ws, choice_from_dict):
```
underscores mean private vars

File: tests/unit/test_install.py [hunk 9]
```diff
@@ -29,10 +34,42 @@
...
 
+# test cluster id
+cluster_id = "9999-999999-abcdefgh"
+
+
+def mock_clusters():
+    from databricks.sdk.service.compute import DataSecurityMode, State
+
+    return [
+        MagicMock(
```
Comment (nfx) [inline tests/unit/test_install.py:50]:
```suggestion
        compute.ClusterInfo(
```

we have a concrete type, use this one

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/install.py (hunk 7 above)
Comment (dmoore247) [reply src/databricks/labs/ucx/install.py]: Done

File: src/databricks/labs/ucx/configure.py (hunk 6 above)
Comment (dmoore247) [reply src/databricks/labs/ucx/configure.py]: Done

File: tests/unit/test_install.py (hunk 9 above)
Comment (dmoore247) [reply tests/unit/test_install.py]: Done

**Review by nfx — APPROVED** [review]

**Pull request merged**